## 1st Assignment 

- Name: Lam Yue Hei, Moon
- Student ID: 21456550


Q1.

In [3]:
from bs4 import BeautifulSoup
import requests 
import pandas as pd
import numpy as np
from time import sleep
from time import time
from random import randint

In [4]:
#pick the language as English 
headers = {"Accept-Language": "en-US, en;q=0.5"}

#pick the url (hiking routes suggested by the government)
url = 'https://www.hiking.gov.hk/trail'
page = requests.get(url, headers = headers)
page


<Response [200]>

Q2(1).

In [5]:
soup = BeautifulSoup(page.content, 'html.parser')

In [6]:
#find the fist title
soup.find('title')

<title>Enjoy Hiking - Hiking Routes</title>

In [7]:
#find the h2
soup.find_all('h2')

[]

Q2(2).

In [8]:
#find the text in the page
page.content

b'<!doctype html>\n\n<html class="no-js en-us font-md hide-hero index" lang="en-us" dir="ltr">\n  <head>\n      <meta charset="utf-8">\n      <meta http-equiv="x-ua-compatible" content="ie=edge">\n      <meta name="viewport" content="width=device-width, initial-scale=1.0">\n      <title>Enjoy Hiking - Hiking Routes</title>\n      <meta name="title" content="Enjoy Hiking - Hiking Routes">\n      <meta name="description" content="Hiking is taking a long walk in the countryside, usually for pleasure or exercise. There are many hiking trails in our Country Parks, and they are all managed by the Country and Marine Parks Authority. To manage hiking trails in remote areas is an extremely demanding task, so you can help by taking your litter home after hiking. To maintain the naturalness of our countryside, we promote the use of natural materials in our trails. Most concrete footpaths to remote villages which are linked with our footpaths are managed by the Home Affairs Department.">\n      <m

In [9]:
trails = soup.find (id = "grid-view")

Q4(1).

In [10]:
#get the element - names of the trails 
names = [day.text for day in trails.find_all(class_="f-sm-1dot25 text-white bg-third px-md-4 px-2 py-2 py-md-3 mb-0 view-info")]
names

['Aberdeen Nature Trail',
 'Aberdeen Tree Walk',
 'Ap Chau Geosite',
 "Bride's Pool Nature Trail",
 'Cheung Sheung Country Trail',
 'Chi Ma Wan Country Trail',
 'Chuen Lung Family Walk',
 'Chung Pui Tree Walk',
 'Clear Water Bay Tree Walk',
 'Double Haven Country Trail',
 "Eagle's Nest Nature Trail",
 'Fan Lau Country Trail',
 'Fu Tei Country Trail',
 'Fung Hang Family Walk',
 'High Island Geo Trail',
 'High Junk Peak Country Trail',
 'Ho Pui Reservoir Family Walk',
 'Hok Tau Country Trail',
 'Hok Tau Reservoir Family Walk',
 'Hong Kong Trail (Section 1)The Peak to Pok Fu Lam Reservoir Road',
 'Hong Kong Trail (Section 2)Pok Fu Lam Reservoir Road to Peel Rise',
 'Hong Kong Trail (Section 3) Peel Rise to Wan Chai Gap',
 'Hong Kong Trail (Section 4)Wan Chai Gap to Wong Nai Chung Gap',
 'Hong Kong Trail (Section 5)Wong Nai Chung Gap to Mount Parker Road',
 'Hong Kong Trail (Section 6) Mount Parker Road to Tai Tam Road ',
 'Hong Kong Trail (Section 7) Tai Tam Road to To Tei Wan',
 'Hong Ko

In [11]:
#get the districts
districts = [sd.text for sd in trails.find_all(class_="text-red px-md-4 px-2 py-1 bg-forth mb-0 region")]
districts

['Hong Kong Island',
 'Hong Kong Island',
 'North New Territories',
 'North New Territories',
 'Sai Kung',
 'Lantau',
 'West New Territories',
 'North New Territories',
 'Sai Kung',
 'North New Territories',
 'Central New Territories',
 'Lantau',
 'West New Territories',
 'North New Territories',
 'Sai Kung',
 'Sai Kung',
 'West New Territories',
 'North New Territories',
 'North New Territories',
 'Hong Kong Island',
 'Hong Kong Island',
 'Hong Kong Island',
 'Hong Kong Island',
 'Hong Kong Island',
 'Hong Kong Island',
 'Hong Kong Island',
 'Hong Kong Island',
 'Hong Kong Island',
 'Central New Territories',
 'Central New Territories',
 'Central New Territories',
 'West New Territories',
 'North New Territories',
 'Sai Kung',
 'Lantau',
 'Sai Kung',
 'North New Territories',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'Lantau',
 'North New Territories',
 'Lantau',
 'Sai Kung',
 'Sai Kung',
 'West New Territ

In [12]:
# get the rating of challenge and love
difficulties = [temp.text for temp in trails.find_all(class_="d-block text-center")]
difficulties

['(1-Star Easy)',
 '(Rating 1-Heart)',
 '(1-Star Easy)',
 '(Rating 4-Heart)',
 '(1-Star Easy)',
 '(Rating 3-Heart)',
 '(1-Star Easy)',
 '(Rating 3-Heart)',
 '(2-Star Moderate)',
 '(Rating 3-Heart)',
 '(4-Star Difficult)',
 '(Rating 4-Heart)',
 '(1-Star Easy)',
 '(Rating 3-Heart)',
 '(1-Star Easy)',
 '(Rating 4-Heart)',
 '(1-Star Easy)',
 '(Rating 3-Heart)',
 '(3-Star Demanding)',
 '(Rating 3-Heart)',
 '(1-Star Easy)',
 '(Rating 3-Heart)',
 '(1-Star Easy)',
 '(Rating 3-Heart)',
 '(2-Star Moderate)',
 '(Rating 3-Heart)',
 '(1-Star Easy)',
 '(Rating 3-Heart)',
 '(2-Star Moderate)',
 '(Rating 3-Heart)',
 '(3-Star Demanding)',
 '(Rating 3-Heart)',
 '(1-Star Easy)',
 '(Rating 2-Heart)',
 '(2-Star Moderate)',
 '(Rating 2-Heart)',
 '(1-Star Easy)',
 '(Rating 3-Heart)',
 '(2-Star Moderate)',
 '(Rating 3-Heart)',
 '(2-Star Moderate)',
 '(Rating 5-Heart)',
 '(2-Star Moderate)',
 '(Rating 3-Heart)',
 '(2-Star Moderate)',
 '(Rating 2-Heart)',
 '(3-Star Demanding)',
 '(Rating 3-Heart)',
 '(2-Star Mo

Q4(2).

In [13]:
#As rating for both challenge and heart are different, 
#this code is to seperate them into two lists by identifying the word "star" in rating difficulties

challenge = []
rating = []

for d in difficulties: 
    if "Star" in d: 
        challenge.append(d)
    else: 
        rating.append(d)

print(challenge)
print(rating)

['(1-Star Easy)', '(1-Star Easy)', '(1-Star Easy)', '(1-Star Easy)', '(2-Star Moderate)', '(4-Star Difficult)', '(1-Star Easy)', '(1-Star Easy)', '(1-Star Easy)', '(3-Star Demanding)', '(1-Star Easy)', '(1-Star Easy)', '(2-Star Moderate)', '(1-Star Easy)', '(2-Star Moderate)', '(3-Star Demanding)', '(1-Star Easy)', '(2-Star Moderate)', '(1-Star Easy)', '(2-Star Moderate)', '(2-Star Moderate)', '(2-Star Moderate)', '(2-Star Moderate)', '(3-Star Demanding)', '(2-Star Moderate)', '(2-Star Moderate)', '(3-Star Demanding)', '(2-Star Moderate)', '(1-Star Easy)', '(1-Star Easy)', '(1-Star Easy)', '(2-Star Moderate)', '(2-Star Moderate)', '(1-Star Easy)', '(3-Star Demanding)', '(1-Star Easy)', '(1-Star Easy)', '(1-Star Easy)', '(4-Star Difficult)', '(4-Star Difficult)', '(2-Star Moderate)', '(3-Star Demanding)', '(2-Star Moderate)', '(3-Star Demanding)', '(2-Star Moderate)', '(2-Star Moderate)', '(2-Star Moderate)', '(2-Star Moderate)', '(3-Star Demanding)', '(3-Star Demanding)', '(2-Star Mode

In [14]:
#find the descriptions of the trails 
descriptions = [sd.text for sd in trails.find_all(class_="summary d-block px-md-2 px-0 pt-2")]
descriptions

['\n                                    The first leg of Aberdeen Nature Trail skirts along Lower Ab...                                ',
 '\n                                    Aberdeen Tree Walk lies near the downtown area and takes abo...                                ',
 '\n                                    Located to the east of Sha Tau Kok, Ap Chau, which literally...                                ',
 '\n                                    Located at the northwestern corner of Plover Cove Country Pa...                                ',
 '\n                                    Cheung Sheung, surrounded by steep spurs, is one of those ra...                                ',
 '\n                                    Chi Ma Wan Country Trail, with a total distance of 18.5 kilo...                                ',
 '\n                                    Situated on a mountain in Tsuen Wan, Chuen Lung is an ancien...                                ',
 '\n                              

Q4(3).

In [16]:
#make sure all elements have the same lens
print(len(names))
print(len(districts))
print(len(challenge))
print(len(rating))
print(len(new))

129
129
129
129
129


In [17]:
#create data frame
places = pd.DataFrame({
 "Names": names,
 "Districts": districts,
 "Difficulties - Over 5 Stars": challenge,
 "Rating - Over 5 Stars": rating,
 "Descriptions": new
})
places

Names              Districts  \
0                      Aberdeen Nature Trail       Hong Kong Island   
1                         Aberdeen Tree Walk       Hong Kong Island   
2                            Ap Chau Geosite  North New Territories   
3                  Bride's Pool Nature Trail  North New Territories   
4                Cheung Sheung Country Trail               Sai Kung   
..                                       ...                    ...   
124                    Wong Shek Family Walk               Sai Kung   
125                      Wong Shek Tree Walk               Sai Kung   
126                Wu Kau Tang Country Trail  North New Territories   
127  Yuen Tsuen Ancient Trail Country Trail    West New Territories   
128                   Yuen Tun Country Trail   West New Territories   

    Difficulties - Over 5 Stars Rating - Over 5 Stars  \
0                 (1-Star Easy)      (Rating 1-Heart)   
1                 (1-Star Easy)      (Rating 4-Heart)   
2                 (1-Star Easy)      (Rating 3-Heart)   
3                 (1-Star Easy)      (Rating 3-Heart)   
4             (2-Star Moderate)      (Rating 3-Heart)   
..                          ...                   ...   
124               (1-Star Easy)      (Rating 3-Heart)   
125               (1-Star Easy)      (Rating 5-Heart)   
126           (2-Star Moderate)      (Rating 3-Heart)   
127          (4-Star Difficult)      (Rating 4-Heart)   
128          (3-Star Demanding)      (Rating 2-Heart)   

                                          Descriptions  
0    The first leg of Aberdeen Nature Trail skirts ...  
1    Aberdeen Tree Walk lies near the downtown area...  
2    Located to the east of Sha Tau Kok, Ap Chau, w...  
3    Located at the northwestern corner of Plover C...  
4    Cheung Sheung, surrounded by steep spurs, is o...  
..                                                 ...  
124  Hailed as the ‘backyard of Hong Kong’, Sai Kun...  
125  Wong Shek Tree Walk is located in Sai Kung Eas...  
126  Feng Shui and numerology play an important rol...  
127  Ancient trails used to be a key route for vill...  
128  The Hakka form an integral part of indigenous ...  

[129 rows x 5 columns]

Q4(3).

In [19]:
#save it to my laptop(with window setting) 
places.to_csv(r'C:\Users\Moon Lam\Desktop\Master\AIDM 7330\places.csv', index = False)

The problem I define is https://www.hiking.gov.hk/trail, a government official website, sharing the
hiking routes and information about hiking in Hong Kong.

In the code, I scrape the webpage in English, finding titles and useful elements. I scrape all the
information about hiking routes. It includes the hiking trail names, districts, difficulties, rating, and
descriptions.

However, the difficulties and rating of trails are in the same class on the webpage. I use the for loop
and conditional statements to separate the elements by identifying the word “stars” as difficulties.
Meanwhile, there are “\n” and several spaces at the beginning of the descriptions. I use for loop
again to beautify the trail descriptions.

I create a data frame that shares the information of the hiking trails with the output as a CSV file.
There are 5 columns (without index) and 129 rows, meaning that there are 129 trails in total. Each
column represents trails’ name, district, difficulties (with five as full marks), ratings (with five as full
marks), and descriptions. Each row means each trail. It is sorted by the trail names in alphabetical
order, the same as the webpage.

Despite that, there are still some bugs in the data frame. First, the number of symbols of hearts or
stars might be different from that in text. For example, there are only three hearts in ratings of
Aberdeen Tree Walk, but it has written “Rating 4-Heart”, reducing the accuracy of the data. Also, the
descriptions of the trails end with “...”, which has not shown the complete description but only a
summary.